hbase_sql/src/main/java/com/splicemachine/mrio/api/core/AbstractSMInputFormat.java

```diff
@@ -191,6 +192,66 @@
         }
     }
 
+    private int compareRows(byte[] left, byte[] right) {
+        return org.apache.hadoop.hbase.util.Bytes.compareTo(left, right);
+    }
+
+    private boolean splitContainsScan(SMSplit split, byte[] scanStartRow, byte[] scanStopRow) {
+        if (scanStartRow.length == 0 && scanStopRow.length == 0) {
+            return compareRows(split.split.getStartRow(), scanStartRow) == 0 &&
+                    compareRows(split.split.getEndRow(), scanStopRow) == 0;
+        } else {
+            if (compareRows(split.split.getStartRow(), scanStartRow) < 1) {
+                if (split.split.getEndRow().length == 0 || compareRows(split.split.getEndRow(), scanStopRow) >= 0) {
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Checks the sequence of split rows. If any gap between splits is found or start/stop row of scan does not
+     * correspond to the generated splits, the splitting has to be recalculated again.
+     *
+     * @param inputSplits generated splits
+     * @param scanStartRow
+     * @param scanStopRow
+     * @return
+     */
+    protected boolean isRefreshNeeded(List<InputSplit> inputSplits, byte[] scanStartRow, byte[] scanStopRow) {
+        assert inputSplits.stream().allMatch(is -> (is instanceof SMSplit)) : "items expected to be instanceof SMSplit";
+
+        if (inputSplits.size() < 2) {
+            return !splitContainsScan((SMSplit) inputSplits.get(0), scanStartRow, scanStopRow);
+        } else {
+            inputSplits.sort(new Comparator<InputSplit>() {
+                @Override
+                public int compare(InputSplit o1, InputSplit o2) {
+                    SMSplit smSplit1 = (SMSplit) o1;
+                    SMSplit smSplit2 = (SMSplit) o2;
+
+                    return compareRows(smSplit1.split.getStartRow(), smSplit2.split.getStartRow());
+                }
+            });
+
+            for (int i = 1; i < inputSplits.size(); i++) {
+                byte currentStartRow[] = ((SMSplit) inputSplits.get(i)).split.getStartRow();
+                byte prevEndRow[] = ((SMSplit) inputSplits.get(i - 1)).split.getEndRow();
+                if (compareRows(currentStartRow, prevEndRow) != 0) {
+                    LOG.warn("The gap in splits is found: current split [" + inputSplits.get(i) + "], previous split [" + inputSplits.get(i - 1) + "]");
+                    return true;
+                }
+            }
+            SMSplit firstSplit = (SMSplit) inputSplits.get(0);
+            SMSplit lastSplit = (SMSplit) inputSplits.get(inputSplits.size() - 1);
+            if (compareRows(firstSplit.split.getStartRow(), scanStartRow) != 0 || compareRows(lastSplit.split.getEndRow(), scanStopRow) != 0) {
```

Comment:
This check is not correct, we can have a scan with start key = 1000 and the first partition might have key = 500.

We have to check that firstSplit.split.getStartRow() <= scanStartRow and that lastSplit.split.getEndRow() >= scanStopRow (or lastSplit.split.getEndRow().isEmpty())

Reply:
it has been corrected